python/src/scipp/coords.py

```diff
@@ -98,60 +120,67 @@ def __init__(self, obj, *, graph, outputs):
         self._outputs = outputs  # names of outputs
         self._graph = graph
 
-    def _add_event_coord(self, key, coord):
-        try:
-            self.obj.bins.coords[key] = coord
-        except VariableError:  # Thrown on mismatching bin indices, e.g. slice
-            self.obj.data = self.obj.data.copy()
-            self.obj.bins.coords[key] = coord
-
-    def _add_event_coords(self, coords):
-        for key, coord in coords.items():
-            # Non-binned coord should be duplicate of dense handling above,
-            # if present => ignored.
-            if coord.bins is not None:
-                self._add_event_coord(key, coord)
-
     def _add_coord(self, *, name):
-        if name in self.obj.meta:
+        if self._exists(name):
             return _produce_coord(self.obj, name)
         if isinstance(self._graph[name], str):
             self._aliases.append(name)
-            out = self._get_coord(self._graph[name])
-            if self.obj.bins is not None:
-                # Calls to _get_coord for dense coord handling take care of
-                # recursion and add also event coords, here and below we thus
-                # simply consume the event coord.
-                if self._graph[name] in self.obj.meta:
-                    out_bins = _consume_coord(self.obj.bins, self._graph[name])
+            out = {name: self._get_coord(self._graph[name])}
             dim = (self._graph[name], )
         else:
             func = self._graph[name]
             argnames = _argnames(func)
             args = {arg: self._get_coord(arg) for arg in argnames}
-            out = func(**args)
-            if self.obj.bins is not None:
-                args.update({
-                    arg: _consume_coord(self.obj.bins, arg)
-                    for arg in argnames if arg in self.obj.bins.meta
-                })
-                out_bins = func(**args)
+            have_all_dense_inputs = all([v[0] is not None for v in args.values()])
+            if have_all_dense_inputs:
+                dense_args = {k: v[0] for k, v in args.items()}
+                out = func(**dense_args)
+                if not isinstance(out, dict):
+                    out = {name: out}
+            else:
+                out = {}
+            have_event_inputs = any([v[1] is not None for v in args.values()])
+            if self.obj.bins is not None and have_event_inputs:
```

Comment:
Is the `self.obj.bins is not None` check required? If `have_event_inputs` is `True`, does that not imply that the object has bins?

Reply:
👍